python/ffsim/states/slater.py

```diff
@@ -0,0 +1,353 @@
+            of the pair contains the 1-RDM for each spin sector.
+        chain_length: Length of each sampling chain.
+        n_chains: Number of chains to run. The total number of samples is given
+            by (``chain_length`` + 1) * ``n_chains``.
+        n_particles_to_move: At each step of the chain, the maximum number of
+            electrons that are moved into a different orbital.
+        seed: Seed for the random number generator.
+
+    Returns:
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+
+    if isinstance(rdm, np.ndarray) and rdm.ndim == 2:
+        # spinless case
+        n = round(np.real(np.sum(np.diag(rdm))))
+        norb = rdm.shape[0]
+
+        if n == 0 or n == norb:
+            sampled_configuration = np.ones(
+                (n_chains * (chain_length + 1), rdm.shape[0]), dtype=int
+            ) * int(n / norb)
+
+        else:
+            sampled_configuration = _sample_spinless(
+                rdm, chain_length, n_chains, n_particles_to_move, seed
+            )
+    else:
+        # Spinful case
+        rdm_a, rdm_b = rdm
+        n_a = round(np.real(np.sum(np.diag(rdm_a))))
+        n_b = round(np.real(np.sum(np.diag(rdm_b))))
+        norb = rdm_a.shape[0]
+
+        if n_a == 0 or n_a == norb:
+            sampled_configuration_a = np.ones(
+                (n_chains * (chain_length + 1), norb), dtype=int
+            ) * int(n_a / norb)
+        else:
+            sampled_configuration_a = _sample_spinless(
+                rdm_a, chain_length, n_chains, n_particles_to_move, seed
+            )
+
+        if n_b == 0 or n_b == norb:
+            sampled_configuration_b = np.ones(
+                (n_chains * (chain_length + 1), norb), dtype=int
+            ) * int(n_a / norb)
+        else:
+            sampled_configuration_b = _sample_spinless(
+                rdm_b, chain_length, n_chains, n_particles_to_move, seed
+            )
+
+        sampled_configuration = np.concatenate(
+            (sampled_configuration_b, sampled_configuration_a), axis=1
+        )
+
+    return sampled_configuration
+
+
+def _propose_move(
+    previous_step: np.ndarray, norb: int, n_particles_to_move: int, rng=None
+) -> np.ndarray:
+    """
+    Proposes a new state in the Markov chain. The new state is obtained by allowing
+    at most ``n_particles_to_move`` to change the orbital that they occupy.
+    The new orbitals that they can occupy are chosen at random from the set of
+    unoccupied orbitals.
+
+    Args:
+        previous step: A Numpy array with the starting positions of the electrons.
+        norb: Number of orbitals.
+        n_particles_to_move: At each step of the chain, the maximum number of
+            electrons that are moved into a different orbital.
+
+    Returns:
+        A Numpy array with the proposed new set of positions of the electrons.
+    """
+
+    rng = np.random.default_rng(rng)
```

Comment:
The `rng` should be initialized at the beginning of `sampler_slater`, and then passed to helper functions.